crates/apollo-compiler/src/database/document.rs

```diff
@@ -38,24 +38,14 @@ pub(crate) fn find_type_definition_by_name(
     db.types_definitions_by_name().get(&name).cloned()
 }
 
-pub(crate) fn find_operation_by_name(
-    db: &dyn HirDatabase,
-    file_id: FileId,
-    name: String,
-) -> Option<Arc<OperationDefinition>> {
-    db.operations(file_id)
-        .iter()
-        .find(|def| def.name() == Some(&*name))
-        .cloned()
-}
-
-pub(crate) fn find_anonymous_operation(
+pub(crate) fn find_operation(
     db: &dyn HirDatabase,
     file_id: FileId,
+    name: Option<String>,
 ) -> Option<Arc<OperationDefinition>> {
     db.operations(file_id)
         .iter()
-        .find(|def| def.name().is_none())
+        .find(|def| def.name() == name.as_deref())
```

Comment:
When the `name` parameter is `None` is looks for an anonymous operation. This deviates from [the spec’s algorithm](https://spec.graphql.org/October2021/#GetOperation()), which instead does something like:


```rust
let operations = db.operations(file_id);
if let Some(name) = name {
    operations.iter().find(|def| def.name() == Some(&*name))
} else if operations.len() == 1 {
    operations[0].clone()
} else {
    None
}

Reply:
oh, good catch, i misread that entirely!

Reply:
we don't (yet, or maybe even never?) return Result from these kinds of queries, so if `None` is provided and there are _several_ operation definitions, this function will return `None`. Does that sound ok to everyone? @SimonSapin @goto-bus-stop?

https://github.com/apollographql/apollo-rs/blob/cd590648ef946333e8cf1c2deec464823d641395/crates/apollo-compiler/src/database/hir_db.rs#L1626-L1630

Reply:
Yes, sounds good. Depending on whether the parameter is `None` (which the caller knows of course) in each case there’s only one error case, so it’s fine IMO to have a `None` return value represent it.